src/components/tooltip/GCMarker.js

```diff
@@ -382,121 +382,155 @@ function _makePhaseTimesArray(
  * sweeping.sweep
  *
- * Then marking.mark_roots, marking.mark_heap.a, sweeping.sweep are the only
- * leaves.  We select these since they will give the person looking at the
- * profile the best clue about which (sub-)phases are taking the longest.
- * For example, it isn't useful to say "sweeping took 200ms" but it is
- * useful to say "sweeping.compacting took 20ms" (if compacting has no
- * sub-phases.
+ * Then marking.mark_roots, marking.mark_heap.objects, sweeping.sweep are the
+ * only original leaves. We will add in "marking.mark_roots (self-time)" and
+ * "marking.mark_heap (self-time)". We select from these leaves since they will
+ * give the person looking at the profile the best clue about which
+ * (sub-)phases are taking the longest.
  *
- * We find leaf phases by constructing a tree of phase times and then
- * reading its leaves.
+ * For example, it isn't useful to say "marking.mark_heap took 200ms" but it is
+ * useful to say "marking.mark_heap.objects took 180ms" (if objects has no
+ * sub-phases). However, if marking.mark_heap took 200ms total yet all of its
+ * sub-phases add up to only 20ms, then we would want to say "marking.mark_heap
+ * (self-time) is 180ms".
  */
 
 type PhaseTreeNode = {|
-  value?: PhaseTimeTuple,
+  value: PhaseTimeTuple,
+  leaf: boolean,
   branches: Map<string, PhaseTreeNode>,
 |};
 
 function _treeInsert(
   tree: Map<string, PhaseTreeNode>,
   path: Array<string>,
   phase: PhaseTimeTuple
-) {
+): void {
   const component = path.shift();
   if (component === undefined) {
     // This path is not a leaf, it can be ignored.
     return;
   }
 
+  const uninitializedPhaseTime = { name: '(temp)', time: 0 };
+
   let node = tree.get(component);
   if (!node) {
     // Make a new node and grow the tree in this direction.
-    node = { branches: new Map() };
-
+    const value = { ...uninitializedPhaseTime };
+    node = { branches: new Map(), leaf: false, value };
     tree.set(component, node);
   }
 
   if (path.length > 0) {
     // There are more path components.  This node should be a branch if it
     // isn't one already.
-    if (node.value) {
-      // We delete the value to change this node from a leaf to a branch.
-      delete node.value;
-    }
+    node.leaf = false;
     _treeInsert(node.branches, path, phase);
   } else {
     // Make the new node leaf node.
-    if (node.value) {
+    if (node.value.name !== uninitializedPhaseTime.name) {
       console.error(
-        'Duplicate phases in _treeInsert in MarkerTooltipContents.js'
+        `Duplicate phase ${phase.name} in _treeInsert in MarkerTooltipContents.js`
       );
       return;
     }
     node.value = phase;
+    node.leaf = true;
   }
 }
 
 function _treeGetLeaves(
   tree: Map<string, PhaseTreeNode>
 ): Array<PhaseTimeTuple> {
   const leaves = [];
-  for (const node of tree.values()) {
-    if (node.value) {
-      leaves.push(node.value);
+  for (const { branches, leaf, value } of tree.values()) {
+    if (leaf) {
```

Comment:
Probably not worth changing, but is this equality always true: `leaf === (branches.size == 0)` ? In that case do we need a leaf boolean?
Again, not worth changing now, unless you want to, I'm just curious.

Reply:
Another possibility is we may want to avoid creating a Map when `leaf` is true, using a Flow union type.

But not worth doing this, because the number of phases is small enough that we don't want to overoptimize :-)

Reply:
> Probably not worth changing, but is this equality always true: `leaf === (branches.size == 0)` ? In that case do we need a leaf boolean?

It is, and we do not.

> Again, not worth changing now, unless you want to, I'm just curious.

I left it that way only because of your comment that being more explicit here was good. :-) Since adding it, I switched things to unconditionally iterating over `branches` without checking `leaf`, and would personally lean towards eliminating `leaf`. But I didn't want to slow down the progress towards landing. I certainly have no objection if you want to kill it.

Reply:
> Another possibility is we may want to avoid creating a Map when `leaf` is true, using a Flow union type.
> 
> But not worth doing this, because the number of phases is small enough that we don't want to overoptimize :-)

As a JS engine developer, I have a stupidly strong tendency to keep objects' shapes (ordered set of own properties) the same. The benefit of that is probably vastly outweighed by the cost of creating dead `Map`s.

But I agree, it doesn't seem worth worrying about either way.